src/serac/physics/contact/contact_data.cpp

```diff
@@ -0,0 +1,294 @@
+ContactData::ContactData(const mfem::ParMesh& mesh)
+    : mesh_{mesh},
+      reference_nodes_{dynamic_cast<const mfem::ParGridFunction*>(mesh.GetNodes())},
+      current_coords_{*reference_nodes_},
+      have_lagrange_multipliers_{false},
+      num_pressure_true_dofs_{0}
+{
+  tribol::initialize(mesh_.SpaceDimension(), mesh_.GetComm());
+}
+
+ContactData::~ContactData() { tribol::finalize(); }
+
+void ContactData::addContactPair(int pair_id, const std::set<int>& bdry_attr_surf1,
+                                 const std::set<int>& bdry_attr_surf2, ContactOptions contact_opts)
+{
+  pairs_.emplace_back(pair_id, mesh_, bdry_attr_surf1, bdry_attr_surf2, current_coords_, contact_opts);
+  if (contact_opts.enforcement == ContactEnforcement::LagrangeMultiplier) {
+    have_lagrange_multipliers_ = true;
+    num_pressure_true_dofs_ += pairs_.back().numPressureTrueDofs();
+  }
+}
+
+bool ContactData::haveContactPairs() const { return !pairs_.empty(); }
+
+void ContactData::update(int cycle, double time, double& dt)
+{
+  tribol::updateMfemParallelDecomposition();
+  tribol::update(cycle, time, dt);
+}
+
+mfem::Vector ContactData::trueContactForces() const
+{
+  mfem::Vector f_true(reference_nodes_->ParFESpace()->GetTrueVSize());
+  mfem::Vector f(reference_nodes_->ParFESpace()->GetVSize());
+  f = 0.0;
+  for (const auto& pair : pairs_) {
+    f += pair.contactForces();
+  }
+  // NOTE: forces are considered a dual field in MFEM and the correct operator
+  // here for MFEM dual fields is P^T.  However, Tribol stores forces similar to
+  // a ParGridFunction -- all shared dofs are equal; therefore, the operator
+  // that returns the desired values is R
+  reference_nodes_->ParFESpace()->GetRestrictionMatrix()->Mult(f, f_true);
+  return f_true;
+}
+
+mfem::Vector ContactData::truePressures() const
+{
+  mfem::Vector p_true(numPressureTrueDofs());
+  auto         dof_offsets = pressureTrueDofOffsets();
+  for (size_t i{0}; i < pairs_.size(); ++i) {
+    if (pairs_[i].getContactOptions().enforcement == ContactEnforcement::LagrangeMultiplier) {
+      mfem::Vector p_pair_true;
+      p_pair_true.MakeRef(p_true, dof_offsets[static_cast<int>(i)],
+                          dof_offsets[static_cast<int>(i) + 1] - dof_offsets[static_cast<int>(i)]);
+      pairs_[i].pressure().ParFESpace()->GetRestrictionMatrix()->Mult(pairs_[i].pressure(), p_pair_true);
+    }
+  }
+  return p_true;
+}
+
+mfem::Vector ContactData::trueGaps() const
+{
+  mfem::Vector g_true(numPressureTrueDofs());
+  auto         dof_offsets = pressureTrueDofOffsets();
+  for (size_t i{0}; i < pairs_.size(); ++i) {
+    if (pairs_[i].getContactOptions().enforcement == ContactEnforcement::LagrangeMultiplier) {
+      mfem::Vector g_pair_true;
+      g_pair_true.MakeRef(g_true, dof_offsets[static_cast<int>(i)],
+                          dof_offsets[static_cast<int>(i) + 1] - dof_offsets[static_cast<int>(i)]);
+      // NOTE: gaps are considered a dual field in MFEM and the correct operator
```

Comment:
@srwopschall , is this the correct operator to use? I vaguely remember you having a similar issue before.